src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3982,14 +3988,176 @@ bool   os::win32::_is_windows_server         = false;
+  if (version_size == 0) {
+    warning("GetFileVersionInfoSize() failed: GetLastError->%ld.", GetLastError());
+    return;
+  }
+
+  LPTSTR version_info = (LPTSTR)os::malloc(version_size, mtInternal);
+  if (version_info == nullptr) {
+    warning("os::malloc() failed to allocate %ld bytes for GetFileVersionInfo buffer", version_size);
+    return;
+  }
+
+  if (!GetFileVersionInfo(kernel32_path, 0, version_size, version_info)) {
+    os::free(version_info);
+    warning("GetFileVersionInfo() failed: GetLastError->%ld.", GetLastError());
+    return;
+  }
+
+  if (!VerQueryValue(version_info, TEXT("\\"), (LPVOID*)&file_info, &len)) {
+    os::free(version_info);
+    warning("VerQueryValue() failed. Cannot determine Windows version.");
+    return;
+  }
+
+  _major_version = HIWORD(file_info->dwProductVersionMS);
+  _minor_version = LOWORD(file_info->dwProductVersionMS);
+  _build_number  = HIWORD(file_info->dwProductVersionLS);
+  _build_minor   = LOWORD(file_info->dwProductVersionLS);
+
+  os::free(version_info);
+}
+
+bool os::win32::is_windows_11_or_greater() {
+  // Windows 11 starts at build 22000 (Version 21H2) as per
+  // https://learn.microsoft.com/en-us/windows/release-health/windows11-release-information
+  return (windows_major_version() >= 10 && windows_build_number() >= 22000 && !IsWindowsServer());
+}
+
+bool os::win32::is_windows_server_2022_or_greater() {
+  // Windows Server 2022 starts at build 20348.169 as per
+  // https://learn.microsoft.com/en-us/windows/release-health/release-information
+  return (windows_major_version() >= 10 && windows_build_number() >= 20348 && IsWindowsServer());
+}
+
+DWORD os::win32::active_processors_in_job_object() {
+  BOOL is_in_job_object = false;
+  if (!IsProcessInJob(GetCurrentProcess(), nullptr, &is_in_job_object)) {
+    warning("IsProcessInJob() failed: GetLastError->%ld.", GetLastError());
+    return 0;
+  }
+
+  if (!is_in_job_object) {
+    return 0;
+  }
+
+  DWORD processors = 0;
+
+  LPVOID job_object_information = nullptr;
+  DWORD job_object_information_length = 0;
+
+  if (!QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, nullptr, 0, &job_object_information_length)) {
+    DWORD last_error = GetLastError();
+    if (last_error == ERROR_INSUFFICIENT_BUFFER) {
+      DWORD group_count = job_object_information_length / sizeof(GROUP_AFFINITY);
+
+      job_object_information = os::malloc(job_object_information_length, mtInternal);
+      if (job_object_information != nullptr) {
+          if (QueryInformationJobObject(nullptr, JobObjectGroupInformationEx, job_object_information, job_object_information_length, &job_object_information_length)) {
+            assert(group_count == job_object_information_length / sizeof(GROUP_AFFINITY), "Unexpected group count");
+
+            for (DWORD i = 0; i < group_count; i++) {
+              KAFFINITY group_affinity = ((GROUP_AFFINITY*)job_object_information)[i].Mask;
```

Comment:
nit: You could create the pointer outside the loop and advance the pointer (rather than dereferencing each time)

Reply:
I think it's easier to read if we're just advancing an index into the loop. I've extracted the base pointer from the loop though.